### DISCUSSION TITLE
How to convert JSON string array to multiple strings

### DISCUSSION THREAD

**Discussion by ptr727** [community/community#26263] [Actions]
<p>Hi, busy creating an action workflow that uses a matrix to build docker containers.<br>
I create JSON to drive the matrix, and some of the attributes are string arrays that will be used for <code>tags:</code> and <code>env:</code> in the docker step.</p>
<p>The <a href="https://github.com/docker/build-push-action/blob/42863b1282f7046f6c1ec9222965bbceafae3bb8/docs/advanced/tags-labels.md" rel="noopener nofollow ugc">build</a> step expects tags, images, environment variables as multi lines, not as an array.</p>
<p>Expected:</p>
<pre><code class="lang-yaml">          tags: |
            type=schedule
            type=ref,event=branch
            type=ref,event=pr
            type=semver,pattern={{version}}
            type=semver,pattern={{major}}.{{minor}}
            type=semver,pattern={{major}}
            type=sha
</code></pre>
<p>How do I convert an array to multiline, or how do I get tags to accept and array?</p>
<pre><code class="lang-json">{
    name: NxMeta-LSIO Latest Develop,
    enable: true,
    push: true,
    context: ./NxMeta-LSIO,
    tags: [
      ptr727/nxmeta-lsio:develop,
      ptr727/nxmeta-lsio:develop-latest,
      ptr727/nxmeta-lsio:develop-5.0.0.35062
    ],
    args: [
      DOWNLOAD_VERSION=5.0.0.35062,
      DOWNLOAD_URL=https://updates.networkoptix.com/metavms/5.0.0.35062%20R9/linux/metavms-server-5.0.0.35062-linux_x64.deb
    ]
  }
</code></pre>
<pre><code class="lang-yaml">- name: 'Build and Push ${{ matrix.containers.name }}'
        uses: docker/build-push-action@v2
        if: ${{ matrix.containers.enable }}
        with:
          context: ${{ matrix.containers.context }}
          file: ${{ matrix.containers.context }}/Dockerfile
          platforms: linux/amd64
          push: ${{ matrix.containers.push }}
          tags: ${{ matrix.containers.tags }}
          build-args: ${{ matrix.containers.args }}
</code></pre>
<p>Error:</p>
<pre><code class="lang-auto">Error: .github/workflows/BuildPublishPipeline.yml (Line: 335, Col: 17): A sequence was not expected
Error: .github/workflows/BuildPublishPipeline.yml (Line: 336, Col: 23): A sequence was not expected
Error: The template is not valid. .github/workflows/BuildPublishPipeline.yml (Line: 335, Col: 17): A sequence was not expected,.github/workflows/BuildPublishPipeline.yml (Line: 336, Col: 23): A sequence was not expected
</code></pre>
<p>Action <a href="https://github.com/ptr727/NxWitness/edit/develop/.github/workflows/BuildPublishPipeline.yml" rel="noopener nofollow ugc">file</a></p>

**Comment by ptr727** [comment]
<p>There is probably a cleaner way, but I did it using jq to flatten the JSON, sed to replace ][, with space, and the encode for inter step transfer.</p>
<pre><code class="lang-yaml">FJSON=$(echo "$JSON" | jq --compact-output)
echo "::set-output name=matrix::${FJSON}"
</code></pre>
<pre><code class="lang-yaml">          Tags="${{ toJson(matrix.containers.tags) }}"
          MLTags=$Tags
          MLTags="$(echo "$MLTags" | sed 's/[][,]/ /g')"
          MLTags="$(echo "$MLTags" | sed '/^[[:space:]]*$/d')"
          ETags=$MLTags
          ETags="${ETags//'%'/'%25'}"
          ETags="${ETags//$'\n'/'%0A'}"
          ETags="${ETags//$'\r'/'%0D'}"
          echo "::set-output name=tags::${ETags}"
</code></pre>
<p>See updated action <a href="https://github.com/ptr727/NxWitness/blob/0d821765db2118b776cfae1a0a3ea6dd7ccda9ac/.github/workflows/BuildPublishPipeline.yml" rel="noopener nofollow ugc">file</a>.</p>
<p>Do please let me know if there is a more elegant solution.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="ptr727" data-post="1" data-topic="260525">
<div class="title">
<div class="quote-controls"></div>
 ptr727:</div>
<blockquote>
<p>How do I convert an array to multiline, or how do I get tags to accept and array?</p>
</blockquote>
</aside>
<p>In principle, the <a href="https://docs.github.com/en/actions/learn-github-actions/expressions#join"><code>join()</code> function</a> should work, though I haven’t tried if <code>'\n'</code> works to make the separator a newline.</p>
<p>Otherwise you could try to build an environment variable or output in a previous step.</p>

**Comment by ptr727** [comment]
<p>Unfortunately it seems that join() with <code>\n</code> does not work, I also tried <code>%0A</code>:</p>
<pre><code class="lang-auto">JSON Args:
[
  DOWNLOAD_VERSION=5.0.0.35062,
  DOWNLOAD_URL=https://updates.networkoptix.com/metavms/5.0.0.35062%20R9/linux/metavms-server-5.0.0.35062-linux_x64.deb
]
Join OA
DOWNLOAD_VERSION=5.0.0.35062%0ADOWNLOAD_URL=https://updates.networkoptix.com/metavms/5.0.0.35062%20R9/linux/metavms-server-5.0.0.35062-linux_x64.deb
Join n
DOWNLOAD_VERSION=5.0.0.35062\nDOWNLOAD_URL=https://updates.networkoptix.com/metavms/5.0.0.35062%20R9/linux/metavms-server-5.0.0.35062-linux_x64.deb
</code></pre>
<p>I tried in the build step in case it is a print issue, but no, seems that <code>\n</code> is interpreted as a literal <code>\</code> and a <code>n</code>:</p>
<pre><code class="lang-auto">Error: buildx failed with: error: invalid tag "***/nxmeta-lsio:develop\\n***/nxmeta-lsio:develop-latest\\n***/nxmeta-lsio:develop-5.0.0.35062": invalid reference format
</code></pre>

**Comment by airtower-luna** [comment]
<p>What if you write the URL escaped version (with <code>%0A</code>) to a step output, and use that output to set the tags, something like <code>tags: ${{ steps.generate_tags.outputs.tags }}</code>?</p>

**Comment by ChristopherHX** [comment, marked as answer]
<blockquote>
<p>How do I convert an array to multiline</p>
</blockquote>
<p>You can use an unescaped newline, you only have to escape <code>'</code> with <code>''</code> in an expression string literal.</p>
<pre><code class="lang-yaml">- name: 'Build and Push ${{ matrix.containers.name }}'
        uses: docker/build-push-action@v2
        if: ${{ matrix.containers.enable }}
        with:
          context: ${{ matrix.containers.context }}
          file: ${{ matrix.containers.context }}/Dockerfile
          platforms: linux/amd64
          push: ${{ matrix.containers.push }}
          tags: |-
            ${{ join(matrix.containers.tags, '
            ') }}
          build-args: ${{ matrix.containers.args }}
</code></pre>
<p>I use <code>|-</code> to prevent a newline appended to the result of the multiline expressions, this is important if the return type is not string.</p>
<p>Here an example workflow</p>
<pre><code class="lang-yaml">on: push
jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        include:
        - l:
          - a
          - b
          - c
          - d
    name: My Job
    steps:
    - run: echo "$ARRAY"
      env:
        ARRAY: |-
          ${{ join(matrix.l, '
          ') }}
</code></pre>

**Comment by ptr727** [comment]
<aside class="quote no-group" data-username="ChristopherHX" data-post="6" data-topic="260525">
<div class="title">
<div class="quote-controls"></div>
 ChristopherHX:</div>
<blockquote>
<p>You can use an unescaped newline, you only have to escape <code>'</code> with <code>''</code> in an expression string literal.</p>
</blockquote>
</aside>
<p>Nice, not obvious, but works, thanks.</p>
